Pq optimized distance between compressed vectors

abdelr opened the pull request:
### What's being changed:

This PR introduces some performance optimisations related to:

- Optimised distance calculation between compressed vectors by using a global distance table
- Memory allocation 
- Size of the final rescored result set
- Prevent insertion of element in the final rescored result set that are not good enough so they would be popped anyways

### Review checklist

- [ ] Documentation has been updated, if necessary. Link to changed documentation:
- [X] Chaos pipeline run or not necessary. Link to pipeline:
- [ ] All new code is covered by tests where it is reasonable.
- [X] Performance tests have been run or not necessary.

abdelr commented:
![output](https://github.com/weaviate/weaviate/assets/7632459/7542f185-5f89-49ac-832a-680d9cf0f988)

parkerduckworth commented on adapters/repos/db/vector/hnsw/search.go:
```diff
@@ -557,22 +557,33 @@ func (h *hnsw) knnSearchByVector(searchVec []float32, k int,
-			res := res.Pop()
-			ids[i] = res.ID
-			i--
+		ids := make([]uint64, 0, res.Len())
+		for _, item := range res.Items() {
+			ids = append(ids, item.ID)
```
Since we are preallocating `ids` with the correct length, it is more efficient to not append to the list, as each append is another allocation. Instead you can populate each index of the preallocated `ids` slice in the loop

aliszka replied:
I believe append in this case will only change len of the slice without allocating new memory.

abdelr replied:
I also think is should only change the size since we reserve the right capacity initially but will not hurt changing it as @parkerduckworth suggested. At the very least we will set length only once.

parkerduckworth commented on adapters/repos/db/vector/hnsw/search.go:
```diff
@@ -557,22 +557,33 @@ func (h *hnsw) knnSearchByVector(searchVec []float32, k int,
-			if res.Len() > ef {
-				res.Pop()
+			if res.Len() == k {
+				if dist < res.Top().Dist {
```
Do we not want to include a result that is == `res.Top().Dist`?

abdelr replied:
Depends on how we break ties. First in wins, then the incoming new vector only gets in if better, not equal. But I would expect ties to be very rare honestly, unless vectors are equal.

parkerduckworth commented on adapters/repos/db/vector/ssdhelpers/product_quantization.go:
```diff
@@ -223,9 +224,23 @@ func NewProductQuantizerWithEncoders(cfg ent.PQConfig, distance distancer.Provid
 	return pq, nil
 }
 
+func (pq *ProductQuantizer) buildGlobalDistances() {
+	pq.globalDistances = make([]float32, pq.ks*pq.ks)
+	for segment := 0; segment < pq.m; segment++ {
+		for i := 0; i < pq.ks; i++ {
+			for j := 0; j < pq.ks; j++ {
+				cX := pq.kms[segment].Centroid(byte(i))
+				cY := pq.kms[segment].Centroid(byte(j))
+				pq.globalDistances[i*pq.ks+j] = pq.distance.Step(cX, cY)
+			}
+		}
+	}
+}
+
```
A comment with an explanation of this formula would be useful here

aliszka commented on adapters/repos/db/vector/ssdhelpers/product_quantization.go:
```diff
@@ -223,9 +224,23 @@ func NewProductQuantizerWithEncoders(cfg ent.PQConfig, distance distancer.Provid
+	for segment := 0; segment < pq.m; segment++ {
+		for i := 0; i < pq.ks; i++ {
+			for j := 0; j < pq.ks; j++ {
+				cX := pq.kms[segment].Centroid(byte(i))
```
this can be moved to outer loop to be called just once per `i` iteration

aliszka commented on adapters/repos/db/vector/ssdhelpers/product_quantization.go:
```diff
@@ -223,9 +224,23 @@ func NewProductQuantizerWithEncoders(cfg ent.PQConfig, distance distancer.Provid
+			for j := 0; j < pq.ks; j++ {
+				cX := pq.kms[segment].Centroid(byte(i))
+				cY := pq.kms[segment].Centroid(byte(j))
+				pq.globalDistances[i*pq.ks+j] = pq.distance.Step(cX, cY)
```
if step(cx, cy) == step(cy, cx) we could performed calculation only for half of the matrix, as it is symmetrical diagonally. So loop could look like (pseudocode)
```
for i = 0; i < max; i++ {
  for j = 0; j <= i; j++ {
    dist[i][j] = dist[j][i] = calculate(i,j)
  }
}
```

abdelr replied:
Good point!

abdelr commented on adapters/repos/db/vector/hnsw/search.go:
```diff
@@ -557,22 +557,33 @@ func (h *hnsw) knnSearchByVector(searchVec []float32, k int,
-			if res.Len() > ef {
-				res.Pop()
+			if res.Len() == k {
+				if dist < res.Top().Dist {
+					res.Insert(id, dist)
+					res.Pop()
+				}
```
Yes, it is the same before we insert. The result set is inverted though, so the top is the worst element actually. If the new element is better than the already worst, we add it and remove the old worst. Makes sense now?

abdelr commented:
I have just addressed the issues from the code review. I think it is ready for merging but I would wait for @parkerduckworth to check if the changes are enough since he requested some minor changes. I will be offline a few hours though so from my side, it looks ready for merging.
